africanus/averaging/baseline_time_and_channel_mapping.py

```diff
@@ -0,0 +1,222 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from collections import namedtuple
+
+import numpy as np
+import numba
+
+import sys
+sys.path.insert(0, '/Users/smasoka/Varsity/codex-africanus/africanus/averaging/')
+from support import unique_time, unique_baselines
+from africanus.util.numba import is_numba_type_none, generated_jit, njit, jit
+
+class RowMapperError(Exception):
+    pass
+
+
+def is_flagged_factory(have_flag_row):
+    if have_flag_row:
+        def impl(flag_row, r):
+            return flag_row[r] != 0
+    else:
+        def impl(flag_row, r):
+            return False
+
+    return njit(nogil=True, cache=True)(impl)
+
+
+def output_factory(have_flag_row):
+    if have_flag_row:
+        def impl(rows, flag_row):
+            return np.zeros(rows, dtype=flag_row.dtype)
+    else:
+        def impl(rows, flag_row):
+            return None
+
+    return njit(nogil=True, cache=True)(impl)
+
+
+def set_flag_row_factory(have_flag_row):
+    if have_flag_row:
+        def impl(flag_row, in_row, out_flag_row, out_row, flagged):
+            if flag_row[in_row] == 0 and flagged:
+                raise RowMapperError("Unflagged input row contributing "
+                                     "to flagged output row. "
+                                     "This should never happen!")
+
+            out_flag_row[out_row] = (1 if flagged else 0)
+    else:
+        def impl(flag_row, in_row, out_flag_row, out_row, flagged):
+            pass
+
+    return njit(nogil=True, cache=True)(impl)
+
+RowMapOutput = namedtuple("RowMapOutput",
+                          ["map", "time", "flag_row"])
+
+@generated_jit(nopython=True, nogil=True, cache=True)
+def baseline_row_mapper(uvw, time, antenna1, antenna2, flag_row=None, bins_for_longest_baseline=None):
+    
+    have_flag_row = not is_numba_type_none(flag_row)
+    is_flagged_fn = is_flagged_factory(have_flag_row)
+
+    output_flag_row = output_factory(have_flag_row)
+    set_flag_row = set_flag_row_factory(have_flag_row)
+    
+    def impl(uvw, time, antenna1, antenna2, flag_row=None, bins_for_longest_baseline=None):
+        
+        bl_uvw_dist = []
+        longest_baseline = np.sqrt((uvw**2).sum(axis=1)).max()
+        
+        # Unique baseline and time 
+        ubl, bl_idx, bl_inv, bl_count = unique_baselines(antenna1, antenna2)
+        utime, _, time_inv, _ = unique_time(time)
+        
+        # Calculate distances of the unique baselines
+        for i in range(ubl.shape[0]): 
+            p,q = ubl[i,0],ubl[i,1]
+            uvwpq = uvw[(antenna1 == p) & (antenna2 == q)] 
+            baseline_dist = np.sqrt((uvwpq**2).sum(axis=1))[0]
```

Comment:
@atemkeng This seems to set the baseline distance to the UVW distance of the first selected row, for the baseline.

Is this accurate and can it be improved upon? /cc @smasoka

Reply:
So would it be more accurate to do?

```python
baseline_dist = np.sqrt((uvwpq**2).sum()
```

Reply:
`baseline_dist = np.abs(uvwpq[:, 0]).sum(axis=0)`
using the u coordinates only